src/terrain/terrain.js

```diff
@@ -274,18 +274,17 @@ export class Terrain extends Elevation {
                         'This leads to lower resolution of hillshade. For full hillshade resolution but higher memory consumption, define another raster DEM source.');
                 }
                 // Lower tile zoom is sufficient for terrain, given the size of terrain grid.
-                const demScale = this.sourceCache.getSource().tileSize / GRID_DIM;
-                const proxyTileSize = this.proxySourceCache.getSource().tileSize;
+                const scaledDemTileSize = this.getScaledDemTileSize();
                 // Dem tile needs to be parent or at least of the same zoom level as proxy tile.
                 // Tile cover roundZoom behavior is set to the same as for proxy (false) in SourceCache.update().
-                this.sourceCache.update(transform, demScale * proxyTileSize, true);
+                this.sourceCache.update(transform, scaledDemTileSize, true);
                 // As a result of update, we get new set of tiles: reset lookup cache.
-                this._findCoveringTileCache[this.sourceCache.id] = {};
+                this.resetTileLookupCache();
             };
 
             if (!this.sourceCache.usedForTerrain) {
                 // Init cache entry.
-                this._findCoveringTileCache[this.sourceCache.id] = {};
+                this.resetTileLookupCache();
```

Comment:
More call to that helper could be replaced [line 301](https://github.com/mapbox/mapbox-gl-js/pull/11276/files#diff-31a4a1f0767190c162e3de763fc3378d79a0f1b509d9e54a626a6fc0a5719580R301), and [420](https://github.com/mapbox/mapbox-gl-js/blob/main/src/terrain/terrain.js#L420) as well.